### ISSUE TITLE
Question/issue: need sequential startup of VMs at boot to avoid race condition 

### ISSUE THREAD

**Issue by tonsimple** [QubesOS/qubes-issues#3735]
### Qubes OS version: 

R 3.2

### Affected component(s): 

Fedora based VMs, Fedora  23, 25, 26

---

### Steps to reproduce the behavior:

Basically, I have a dedicated VPN VM and a fedora-minimal proxy-VM that is attached to it so that everything that connects to the proxyvm goes through VPN.

All VMs (the proxyvm, the vpnvm, and the vpns that connect to them) are set to autostart on boot

approximately 19 reboots out of 20 everything works "as expected"

about 1 reboot out of 20 the proxyvm just can't send a packet out

This suggests a race condition that I can't quite catch (and rebooting the machine constantly is kind of annoying)

### Expected behavior:

the config always works like it does most of the time

### Actual behavior:

about 1 reboot out of 20 the proxyvm just can't send a packet out

### General notes:

When I start manually the issue only arises (about 1 trial out of 40 or so) when I shut them all down, then try to bring up a VM connected to proxyvm (thus qubes try to bring everything up automatically)

However, if I bring them up manually in sequence, the issue *never* arises (>100 attempts)

Thus, the easiest fix would be to have Qubes bring them up slowly, one by one, in order of 

[start up the main net vm]
_[start up completes]_
[start up vpnvm]
_[start up completes]_
[start up proxyvm]
_[start up completes]_
[start up other qubes else]

However, I can't for the life of mine figure out how to edit VM startup order (ideally, I'd like to have a small pause between them as well, just in case)

---

**Comment by alimirjamali** [comment]
This old issue is still very relevant.

I believe concurrent qube startup should be a queue of `QubesHost.no_cpus() * 2` size. Until they become responsive. Or timeout happens. Then the next ones in queue should be proceed.

Similarly for shutdowns.

**Comment by alimirjamali** [comment]
Related: #5631

**Comment by Harshit6-b** [comment]
Hi, I'd like to work on this issue. Before writing any code I want to confirm my approach with the team.
My analysis: the race happens because each `qubes-vm@<name>.service` unit fires independently with no ordering between them. The `start()` method in `qubesvm.py` already handles recursive netvm starting correctly — the problem is purely at the systemd level.
Proposed fix: in `qubesvm.py`, when a VM's `autostart` is set to True and its `netvm` also has `autostart=True`, write a systemd drop-in at `/etc/systemd/system/qubes-vm@<child>.service.d/50-netvm-after.conf` containing:
```
[Unit]
After=qubes-vm@<netvm>.service
Requires=qubes-vm@<netvm>.service
```
This drop-in gets written/updated in `on_property_pre_set_autostart` and also when `netvm` changes. No new files, no new services — just systemd doing what it was built for.
Does this approach align with what the team has in mind? Happy to adjust before writing code.

**Comment by marmarek** [comment]
No, I don't think it would change anything, proxyvm is already started before in the `start()` method, as you noticed.
It's more likely something inside such proxyvm that fails at some times, for example when another qube connects to it early. Maybe it's similar issue to #10643 ?

**Comment by Harshit6-b** [comment]
After looking at #10643 and tracing the startup sequence, when another qube connects to the proxyvm early, a SIGHUP is sent to `qubes-firewall` telling it to reload. The signal handler sets `reload_requested = True`. But if this happens while `update_connected_ips()` is still running during initialization, the SIGHUP interrupts `qdb.read()` inside `get_connected_ips()` which has no retry logic unlike `read_watch()` and the process crashes. I'm thinking of fixing this by adding retry logic in `get_connected_ips()` on `InterruptedError`, similar to how `read_watch()` already handles it. Does this match what you had in mind?

**Comment by marmarek** [comment]
Yes, I think this should should work. Alternative would be to block SIGINT while performing an update, both should work.

**Comment by Harshit6-b** [comment]
Submitted a PR: https://github.com/QubesOS/qubes-core-agent-linux/pull/644
Added a retry loop in `get_connected_ips()` on `InterruptedError`, matching the existing pattern in `read_watch()`. Also added a test that simulates `InterruptedError` on `qdb.read()` — it fails without the fix and passes with it.

**Comment by ben-grande** [comment]
The PR commit from the PR was picked but the PR was not merged: https://github.com/QubesOS/qubes-core-agent-linux/pull/644#issuecomment-4158669108. Can anyone that has encountered this in the past, test if it was fixed?
